src/main/java/org/spongepowered/api/resourcepack/ResourcePack.java

```diff
@@ -0,0 +1,43 @@
+package org.spongepowered.api.resourcepack;
+
+import com.google.common.base.Optional;
+
+import java.net.URL;
+
+public interface ResourcePack {
+
+    /**
+     * Gets the URL associated with this ResourcePack.
+     * 
+     * @return The URL associated with this ResourcePack
+     */
+    URL getURL();
+
+    /**
+     * Gets the name of this resource pack. This is the filename of the pack
+     * zipfile, with all non-word characters removed. Note to implementers: The
+     * name <strong>MUST</strong> be calculated like above, or this API will not
+     * work.
+     * 
+     * @return The name of this resource pack. This is the filename of the pack
+     *         .zip, with all non-word characters removed
+     */
+    String getName();
+
+    /**
+     * Gets the ID of this resource pack.
+     * 
+     * @return the ID of this resource pack
+     */
+    String getId();
+
+    /**
+     * If this resource pack was initialized through
+     * {@link ResourcePackFactory#fromURL(URL)}, the hash, as calculated with
```

Comment:
@ST-DDT The hash generation must work like vanilla or the API will not work with vanilla clients.
@Aaron1011 I know, but if it's a valid SHA1 hash, the client uses it, and if it's invalid the client uses it's own method, as described in getName().